From 982d737e950bb2e44885b26776035bc453415075 Mon Sep 17 00:00:00 2001
From: Antonio Borneo <antonio.borneo@foss.st.com>
Date: Wed, 5 Nov 2025 17:43:09 +0100
Subject: [PATCH] pmdomain: st: simplify STM32MP power domain

Drop the 'sync_state' requirement in the driver.
This makes superfluous other code that is dropped too.

Signed-off-by: Antonio Borneo <antonio.borneo@foss.st.com>
Change-Id: I7442d8ff6ae85da71fe092c120704d08f840b9d0
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/506486
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
---
 drivers/pmdomain/st/stm32mp-pm-domain.c | 41 ++-----------------------
 1 file changed, 2 insertions(+), 39 deletions(-)

--- a/drivers/pmdomain/st/stm32mp-pm-domain.c
+++ b/drivers/pmdomain/st/stm32mp-pm-domain.c
@@ -2,7 +2,7 @@
 /*
  * Implements PM domain using the generic PM domain for STM32MP SoC.
  *
- * Copyright (C) STMicroelectronics 2024
+ * Copyright (C) STMicroelectronics 2024-2025
  * Author(s): Patrick Delaunay <patrick.delaunay@foss.st.com>
  */
 #include <linux/device.h>
@@ -14,43 +14,12 @@
 #include <linux/platform_device.h>
 #include <linux/printk.h>
 #include <linux/slab.h>
-#include <linux/suspend.h>
-
-static bool stm32mp_pm_domain_state;
 
 static const struct of_device_id stm32mp_pm_domain_matches[] = {
 	{ .compatible = "st,stm32mp-pm-domain", },
 	{ },
 };
 
-static void stm32mp_pm_domain_sync_state(struct device *dev)
-{
-	/*
-	 * All devices have now been attached/probed to the PM domain topology,
-	 * hence it's fine to allow domain states to be picked.
-	 */
-	stm32mp_pm_domain_state = true;
-}
-
-static int stm32mp_pm_domain_power_off(struct generic_pm_domain *pd)
-{
-	if (!stm32mp_pm_domain_state)
-		return -EBUSY;
-
-	return 0;
-}
-
-static int __maybe_unused stm32mp_pm_domain_suspend(struct device *dev)
-{
-	bool s2idle = (pm_suspend_target_state == PM_SUSPEND_TO_IDLE);
-
-	/* S2IDLE is not allowed when the domain state is unknown */
-	if (s2idle && !stm32mp_pm_domain_state)
-		return -EBUSY;
-
-	return 0;
-}
-
 static void stm32mp_pm_domain_genpd_remove(void *data)
 {
 	pm_genpd_remove(data);
@@ -70,6 +39,7 @@ static int stm32mp_pm_domain_probe(struc
 		pr_warn("%pOF lacks #power-domain-cells\n", np);
 		return 0;
 	}
+
 	pd = kzalloc(sizeof(*pd), GFP_KERNEL);
 	if (!pd)
 		return -ENOMEM;
@@ -81,7 +51,6 @@ static int stm32mp_pm_domain_probe(struc
 	 * support, the wake-up is handled by power domain of interrupt driver
 	 */
 	pd->flags = GENPD_FLAG_PM_CLK | GENPD_FLAG_IRQ_SAFE;
-	pd->power_off = stm32mp_pm_domain_power_off;
 
 	ret = pm_genpd_init(pd, NULL, false);
 	if (ret)
@@ -113,16 +82,10 @@ static int stm32mp_pm_domain_remove(stru
 	return 0;
 }
 
-static const struct dev_pm_ops stm32mp_pm_domain_pm_ops = {
-	SET_SYSTEM_SLEEP_PM_OPS(stm32mp_pm_domain_suspend, NULL)
-};
-
 static struct platform_driver stm32mp_pm_domain_driver = {
 	.driver = {
 		.name = "stm32mp_pm_domain",
 		.of_match_table = stm32mp_pm_domain_matches,
-		.sync_state = stm32mp_pm_domain_sync_state,
-		.pm = &stm32mp_pm_domain_pm_ops,
 	},
 	.probe  = stm32mp_pm_domain_probe,
 	.remove  = stm32mp_pm_domain_remove,
